src/util/sentinel.hpp

```diff
@@ -163,6 +161,16 @@ class sentinel_iterator {
     bool operator>(const sentinel_iterator& x) const {
         return !(x<=*this);
     }
+
+    // access to underlying iterator/sentinel
+
+    bool is_sentinel() const { return e_.index()!=0; }
+    bool is_iterator() const { return e_.index()==0; }
```

Comment:
would it be better to define one of the above, and write the other in terms of it? Ignore this if you can justify your actions!